src/libs/Violations/ViolationsUtils.ts

```diff
@@ -49,31 +49,55 @@ function getTagViolationsForSingleLevelTags(
 }
 
 /**
- * Calculates some tag levels required and missing tag violations for the given transaction
+ * Calculates missing tag violations for policies with dependent tags,
+ * by returning one per tag with its corresponding tagName in the data
  */
-function getTagViolationsForMultiLevelTags(
-    updatedTransaction: Transaction,
-    transactionViolations: TransactionViolation[],
-    policyRequiresTags: boolean,
-    policyTagList: PolicyTagList,
-): TransactionViolation[] {
+function getTagViolationsForDependentTags(policyTagList: PolicyTagList, transactionViolations: TransactionViolation[], tagName: string | undefined) {
+    const tagViolations = [...transactionViolations];
+
+    if (!tagName) {
+        Object.values(policyTagList).forEach((tagList) =>
+            tagViolations.push({
+                name: CONST.VIOLATIONS.MISSING_TAG,
+                type: CONST.VIOLATION_TYPES.VIOLATION,
+                data: {tagName: tagList.name},
+            }),
+        );
+    } else {
+        const tags = TransactionUtils.getTagArrayFromName(tagName);
+        if (Object.keys(policyTagList).length !== tags.length || tags.includes('')) {
+            tagViolations.push({
+                name: CONST.VIOLATIONS.ALL_TAG_LEVELS_REQUIRED,
+                type: CONST.VIOLATION_TYPES.VIOLATION,
+                data: undefined,
+            });
+        }
+    }
+
+    return tagViolations;
+}
+
+/**
+ * Calculates missing tag violations for policies with independent tags,
+ * by returning one per tag with its corresponding tagName in the data
+ */
+function getTagViolationForIndependentTags(policyTagList: PolicyTagList, transactionViolations: TransactionViolation[], transaction: Transaction) {
     const policyTagKeys = getSortedTagKeys(policyTagList);
-    const selectedTags = updatedTransaction.tag?.split(CONST.COLON) ?? [];
+    const selectedTags = transaction.tag?.split(CONST.COLON) ?? [];
     let newTransactionViolations = [...transactionViolations];
-    newTransactionViolations = newTransactionViolations.filter(
-        (violation) => violation.name !== CONST.VIOLATIONS.SOME_TAG_LEVELS_REQUIRED && violation.name !== CONST.VIOLATIONS.TAG_OUT_OF_POLICY,
-    );
 
     // We first get the errorIndexes for someTagLevelsRequired. If it's not empty, we puth SOME_TAG_LEVELS_REQUIRED in Onyx.
     // Otherwise, we put TAG_OUT_OF_POLICY in Onyx (when applicable)
     const errorIndexes = [];
+
```

Comment:
All of these added whitespaces should be removed. Our style guide is to not have blank lines like these unless they're preceeded by comments

Reply:
@cead22 I've never seen that rule before, could you show it to me? Usually I add these white lines to make the code easier to read 😅

Reply:
It's in our [php style guide](https://github.com/Expensify/Style-Guides/blob/main/php.md#blank-lines), but imo it should be in all of them so we're consistent. I'm also of the opinion that these can help readability, but since this is subjective and considered a personal preference, the idea is that we follow the guideline and not making decisions about which style is better, individually, on each of our PRs.

This isn't a criticism, I'm just sharing how I think about this

You can consider these non-blockers

Reply:
I agree with you that there should be consistency throughout the codebase! Maybe you can open an issue and propose these changes to the app style guidelines and let me know 😄